packages/devtools-proxy-support/src/agent.ts

```diff
@@ -0,0 +1,116 @@
+import type { ConnectionOptions } from 'tls';
+import type { Duplex } from 'stream';
+import { SSHAgent } from './ssh';
+import type { ProxyLogEmitter } from './logging';
+import type { EventEmitter } from 'events';
+import type { AgentConnectOpts } from 'agent-base';
+
+// Helper type that represents an https.Agent (= connection factory)
+// with some custom properties that TS does not know about and/or
+// that we add for our own purposes.
+export type AgentWithInitialize = Agent & {
+  // This is genuinely custom for our usage (to allow establishing an SSH tunnel
+  // first before starting to push connections through it)
+  initialize?(): Promise<void>;
+  logger?: ProxyLogEmitter;
+  readonly proxyOptions?: Readonly<DevtoolsProxyOptions>;
+
+  // This is just part of the regular Agent interface, used by Node.js itself,
+  // but missing from @types/node
+  createSocket(
+    req: ClientRequest,
+    options: TcpNetConnectOpts | ConnectionOptions,
+    cb: (err: Error | null, s?: Duplex) => void
+  ): void;
+
+  // http.Agent is an EventEmitter, just missing from @types/node
+} & Partial<EventEmitter>;
+
+class DevtoolsProxyAgent extends ProxyAgent implements AgentWithInitialize {
+  readonly proxyOptions: DevtoolsProxyOptions;
+  private sshAgent: SSHAgent | undefined;
+
+  // Store the current ClientRequest for the time between connect() first
+  // being called and the corresponding _getProxyForUrl() being called.
+  // In practice, this is instantaneous, but that is not guaranteed by
+  // the `ProxyAgent` API contract.
+  // We use a Promise lock/mutex to avoid concurrent accesses.
+  private _req: ClientRequest | undefined;
+  private _reqLock: Promise<void> | undefined;
+  private _reqLockResolve: (() => void) | undefined;
+
+  constructor(proxyOptions: DevtoolsProxyOptions) {
+    super({
+      getProxyForUrl: (url: string) => this._getProxyForUrl(url),
+      ...proxyOptions,
+    });
+    this.proxyOptions = proxyOptions;
+    // This could be made a bit more flexible by actually dynamically picking
+    // ssh vs. other proxy protocols as part of connect(), if we want that at some point.
+    if (proxyOptions.proxy && new URL(proxyOptions.proxy).protocol === 'ssh:') {
+      this.sshAgent = new SSHAgent(proxyOptions);
+    }
+  }
+
+  _getProxyForUrl = (url: string): string => {
+    if (!this._reqLockResolve || !this._req) {
+      throw new Error('getProxyForUrl() called without pending request');
+    }
+    this._reqLockResolve();
+    const req = this._req;
+    this._req = undefined;
+    this._reqLock = undefined;
+    this._reqLockResolve = undefined;
+    return proxyForUrl(this.proxyOptions, url, req);
+  };
+
+  async initialize(): Promise<void> {
+    await this.sshAgent?.initialize();
+  }
+
+  override async connect(
+    req: ClientRequest,
+    opts: AgentConnectOpts
+  ): Promise<HTTPAgent> {
+    if (this.sshAgent) return this.sshAgent;
+    while (this._reqLock) {
+      await this._reqLock;
+    }
```

Comment:
Just curious why we are using while here: is there a possibility of a microtask happening between await resolving and the execution going to the beginning of the loop to check the `_reqLock`?

Reply:
> is there a possibility of a microtask happening between await resolving and the execution going to the beginning of the loop to check the `_reqLock`?

The `await` here does always wait at least one microtask queue iteration, and during that iteration other things can happen (such as other Promises being resolved and handlers for that resolution running)

> Just curious why we are using while here

Yeah, it's to avoid a scenario where if `.connect()` gets called twice (or more) in a synchronous code block, both of those `.connect()` calls would await the same `_reqLock` Promise and then simultaneously proceed (and each of them create new, independent `_reqLock` promises, overwriting each other)

Reply:
Aha, got it, thanks for a detailed explanation with an example, it like made sense, but was still hard to wrap my head around the details of the flow

Reply:
Yeah ... it didn't feel worth pulling in an external library for this, but if you feel like that would be a good idea to make it more obvious what's happening here, I am not opposed to that at all (because it's not just you, I also had to give this a bit of thought to be confident that it's correct the way it is)

Reply:
Sure, added a small comment about this as well 👍